parser/prism/Translator.cc

```diff
@@ -0,0 +1,2011 @@
+                return make_unique<parser::Args>(location, NodeVec{});
+            }
+
+            // Sorbet's legacy parser inserts locals (Shadowargs) into the block's Args node, along with its other
+            // parameters. So we need to extract the args vector from the Args node, and insert the locals at the end of
+            // it.
+            auto sorbetArgsNode = translate(up_cast(paramsNode->parameters));
+            auto argsNode = dynamic_cast<parser::Args *>(sorbetArgsNode.get());
+            auto sorbetShadowArgs = translateMulti(paramsNode->locals);
+            // Sorbet's legacy parser inserts locals (Shadowargs) at the end of the the block's Args node
+            argsNode->args.insert(argsNode->args.end(), std::make_move_iterator(sorbetShadowArgs.begin()),
+                                  std::make_move_iterator(sorbetShadowArgs.end()));
+
+            return sorbetArgsNode;
+        }
+        case PM_BREAK_NODE: { // A `break` statement, e.g. `break`, `break 1, 2, 3`
+            auto breakNode = down_cast<pm_break_node>(node);
+
+            auto arguments = translateArguments(breakNode->arguments);
+
+            return make_unique<parser::Break>(location, move(arguments));
+        }
+        case PM_CALL_AND_WRITE_NODE: { // And-assignment to a method call, e.g. `a.b &&= false`
+            auto flags = static_cast<pm_call_node_flags>(node->flags);
+
+            if (flags & PM_CALL_NODE_FLAGS_SAFE_NAVIGATION) {
+                return translateOpAssignment<pm_call_and_write_node, parser::AndAsgn, parser::CSend>(node);
+            }
+
+            return translateOpAssignment<pm_call_and_write_node, parser::AndAsgn, parser::Send>(node);
+        }
+        case PM_CALL_NODE: { // A method call like `a.b()` or `a&.b()`
+            auto callNode = down_cast<pm_call_node>(node);
+
+            auto loc = location;
+
+            auto name = parser.resolveConstant(callNode->name);
+            auto receiver = translate(callNode->receiver);
+
+            core::LocOffsets messageLoc;
+
+            // When the message is empty, like `foo.()`, the message location is the same as the call operator location
+            if (callNode->message_loc.start == nullptr && callNode->message_loc.end == nullptr) {
+                messageLoc = translateLoc(callNode->call_operator_loc);
+            } else {
+                messageLoc = translateLoc(callNode->message_loc);
+            }
+
+            // Handle `~[Integer]`, like `~42`
+            // Unlike `-[Integer]`, Prism treats `~[Integer]` as a method call
+            // But Sorbet's legacy parser treats both `~[Integer]` and `-[Integer]` as integer literals
+            if (name == "~" && dynamic_cast<parser::Integer *>(receiver.get())) {
+                std::string valueString(reinterpret_cast<const char *>(callNode->base.location.start),
+                                        callNode->base.location.end - callNode->base.location.start);
```

Comment:
This kind of code would be another candidate for the `sliceLocation` helper mentioned elsewhere.